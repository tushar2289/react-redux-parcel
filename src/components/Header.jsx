import React from 'react';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import AppBar from '@material-ui/core/AppBar';
import IconButton from '@material-ui/core/IconButton';
import MenuIcon from '@material-ui/icons/Menu';
import Toolbar from '@material-ui/core/Toolbar';
import { makeStyles } from '@material-ui/core/styles';
import Tooltip from '@material-ui/core/Tooltip';
import Typography from '@material-ui/core/Typography';
import Grid from '@material-ui/core/Grid';
import LocationSearchingIcon from '@material-ui/icons/LocationSearching';
import { drawerWidth } from './Drawer';
import Snackbar from '@material-ui/core/Snackbar';
import MuiAlert from '@material-ui/lab/Alert';
import { setCurrentLocation } from '../features/currentlocation/slice';

function Alert(props) {
  return <MuiAlert elevation={6} variant="filled" {...props} />;
}

const useStyles = makeStyles((theme) => ({
  appBar: {
    [theme.breakpoints.up('sm')]: {
      width: `calc(100% - ${drawerWidth}px)`,
      marginLeft: drawerWidth,
    },
  },
  menuButton: {
    marginRight: theme.spacing(2),
    [theme.breakpoints.up('sm')]: {
      display: 'none',
    },
  },
}));

const Header = ({ handleDrawerToggle }) => {
  const dispatch = useDispatch();

  const getCurrentLocation = () => {
    if (!window.navigator.geolocation) {
      handleClick();
    } else {
      navigator.geolocation.getCurrentPosition(
        (position) => {
          const lat = position.coords.latitude;
          const lng = position.coords.longitude;
          dispatch(setCurrentLocation({ lat, lng }));
        },
        () => {
          handleClick();
        }
      );
    }
  };

  const [open, setOpen] = React.useState(false);

  const handleClick = () => {
    setOpen(true);
  };

  const handleClose = (event, reason) => {
    if (reason === 'clickaway') {
      return;
    }

    setOpen(false);
  };

  const classes = useStyles();
  return (
    <AppBar position="fixed" className={classes.appBar}>
      <Toolbar>
        <Grid justify="space-between" alignContent="center" container>
          <Grid item>
            <IconButton
              color="inherit"
              aria-label="open drawer"
              edge="start"
              onClick={handleDrawerToggle}
              className={classes.menuButton}
            >
              <MenuIcon />
            </IconButton>
            <Typography variant="h6" noWrap style={{ marginTop: '6px' }}>
              Places Nearby
            </Typography>
          </Grid>
          <Grid item>
            <Tooltip title="Get current location">
              <IconButton
                aria-label="Get current location"
                color="inherit"
                onClick={getCurrentLocation}
                data-testid="get_location_button"
              >
                <LocationSearchingIcon />
              </IconButton>
            </Tooltip>
          </Grid>
        </Grid>
      </Toolbar>
      <Snackbar open={open} autoHideDuration={6000} onClose={handleClose}>
        <Alert onClose={handleClose} severity="error">
          Please provide permission to capture location
        </Alert>
      </Snackbar>
    </AppBar>
  );
};

Header.propTypes = {
  handleDrawerToggle: PropTypes.func.isRequired,
};

export default Header;
