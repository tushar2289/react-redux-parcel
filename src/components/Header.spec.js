import React from 'react';
import { Provider } from 'react-redux';
import { createMount } from '@material-ui/core/test-utils';
import store from '../__mocks__/mockStore';
import Header from './Header';

describe('App component', () => {
  let shallow, wrapper;
  beforeEach(() => {
    shallow = createMount();
    wrapper = shallow(
      <Provider store={store}>
        <Header />
      </Provider>
    );
  });

  afterEach(() => {
    shallow = null;
    wrapper = null;
  });

  it('should render properly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should not update location of user since geo location isnt mocked', () => {
    const button = wrapper.find('[data-testid="get_location_button"]');
    button.first().simulate('click');
    expect(store.getState().currentLocation).toMatchSnapshot();
  });

  it('should update location', () => {
    const mockGeolocation = {
      getCurrentPosition: jest.fn().mockImplementation((callback) => {
        const position = {
          coords: {
            latitude: 10,
            longitude: 10,
          },
        };
        callback(position);
      }),
      watchPosition: jest.fn(),
    };

    global.navigator.geolocation = mockGeolocation;

    const button = wrapper.find('[data-testid="get_location_button"]');
    button.first().simulate('click');
    expect(store.getState().currentLocation).toMatchSnapshot();
  });
});
