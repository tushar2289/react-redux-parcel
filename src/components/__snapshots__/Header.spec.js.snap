// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should not update location of user since geo location isnt mocked 1`] = `
Object {
  "lat": 0,
  "lng": 0,
}
`;

exports[`App component should render properly 1`] = `"<header class=\\"MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorPrimary makeStyles-appBar-1 mui-fixed MuiPaper-elevation4\\"><div class=\\"MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters\\"><div class=\\"MuiGrid-root MuiGrid-container MuiGrid-align-content-xs-center MuiGrid-justify-xs-space-between\\"><div class=\\"MuiGrid-root MuiGrid-item\\"><button class=\\"MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-2 MuiIconButton-colorInherit MuiIconButton-edgeStart\\" tabindex=\\"0\\" type=\\"button\\" aria-label=\\"open drawer\\"><span class=\\"MuiIconButton-label\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z\\"></path></svg></span><span class=\\"MuiTouchRipple-root\\"></span></button><h6 class=\\"MuiTypography-root MuiTypography-h6 MuiTypography-noWrap\\" style=\\"margin-top: 6px;\\">Places Nearby</h6></div><div class=\\"MuiGrid-root MuiGrid-item\\"><button class=\\"MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit\\" tabindex=\\"0\\" type=\\"button\\" aria-label=\\"Get current location\\" data-testid=\\"get_location_button\\" title=\\"Get current location\\"><span class=\\"MuiIconButton-label\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M20.94 11c-.46-4.17-3.77-7.48-7.94-7.94V1h-2v2.06C6.83 3.52 3.52 6.83 3.06 11H1v2h2.06c.46 4.17 3.77 7.48 7.94 7.94V23h2v-2.06c4.17-.46 7.48-3.77 7.94-7.94H23v-2h-2.06zM12 19c-3.87 0-7-3.13-7-7s3.13-7 7-7 7 3.13 7 7-3.13 7-7 7z\\"></path></svg></span><span class=\\"MuiTouchRipple-root\\"></span></button></div></div></div></header>"`;

exports[`App component should update location 1`] = `
Object {
  "lat": 10,
  "lng": 10,
}
`;
